graphs.py

```diff
@@ -2,15 +2,44 @@
 from collections import defaultdict
 
 def make_complete_graph(n):
-    return make_random_graph(n, 1)
+    g = defaultdict(list)
+    for i in range(n):
+        for j in range(n):
+            if i == j:
+                continue
+            g[i].append(j)
+            g[j].append(i)
+    return g
 
 def make_random_graph(n, density):
-    g = defaultdict(list)
+    # we want the graph to be connected, so we start with a spanning tree
+    g = make_random_st(n)
     for i in range(n):
         for j in range(n):
```

Comment:
there seems to be a slight problem now though, which is that now we might be adding duplicate edges. If i=0 and j=1, we might add edges (0,1) and (1,0), and later when i=1 and j=0, might add (1, 0) and (0, 1) again. Need to insert the check for i < j i think.

Reply:
Does the check in L21 handle this? 

I think it will:
i = 0,  j = 1 -> g = { 0: [1], 1: [0] }
i = 1,   j= 0 -> 0 is in g = [1], continue

Reply:
oh yes you are right! Sorry about that!